File: java/tutorial/main/ch/softappeal/yass/tutorial/HelloWorld.java
package ch.softappeal.yass.tutorial;

import ch.softappeal.yass.remote.ContractId;
import ch.softappeal.yass.remote.Server;
import ch.softappeal.yass.remote.Service;
import ch.softappeal.yass.remote.SimpleMethodMapper;
import ch.softappeal.yass.serialize.JavaSerializer;
import ch.softappeal.yass.transport.socket.SimpleSocketTransport;
import ch.softappeal.yass.transport.socket.SocketBinder;
import ch.softappeal.yass.transport.socket.SocketConnector;

import java.net.InetSocketAddress;
import java.util.concurrent.Executors;

public class HelloWorld {

    public interface Calculator {
        int add(int a, int b);
    }

    static class CalculatorImpl implements Calculator {
        @Override public int add(int a, int b) {
            return a + b;
        }
    }

    public static void main(String... args) {

        var address = new InetSocketAddress("localhost", 28947);
        var serializer = JavaSerializer.INSTANCE;
        var calculatorId = ContractId.create(Calculator.class, 0, SimpleMethodMapper.FACTORY);

        // start server
        var server = new Server(new Service(calculatorId, new CalculatorImpl()));
        var executor = Executors.newCachedThreadPool();
        new SimpleSocketTransport(executor, serializer, server).start(executor, SocketBinder.create(address));

        // use client
        var client = SimpleSocketTransport.client(serializer, SocketConnector.create(address));
        var calculator = client.proxy(calculatorId);
        System.out.println("2 + 3 = " + calculator.add(2, 3));

    }

}

File: kotlin/tutorial/main/ch/softappeal/yass/tutorial/HelloWorld.kt
package ch.softappeal.yass.tutorial

import ch.softappeal.yass.remote.Server
import ch.softappeal.yass.remote.Service
import ch.softappeal.yass.remote.SimpleMethodMapperFactory
import ch.softappeal.yass.remote.contractId
import ch.softappeal.yass.serialize.JavaSerializer
import ch.softappeal.yass.transport.ClientSetup
import ch.softappeal.yass.transport.ServerSetup
import ch.softappeal.yass.transport.messageSerializer
import ch.softappeal.yass.transport.socket.socketBinder
import ch.softappeal.yass.transport.socket.socketClient
import ch.softappeal.yass.transport.socket.socketConnector
import ch.softappeal.yass.transport.socket.socketServer
import java.net.InetSocketAddress
import java.util.concurrent.Executors

interface Calculator {
    fun add(a: Int, b: Int): Int
}

class CalculatorImpl : Calculator {
    override fun add(a: Int, b: Int) = a + b
}

fun useCalculator(calculator: Calculator) {
    println("2 + 3 = " + calculator.add(2, 3))
}

fun main(args: Array<String>) {
    val calculatorId = contractId<Calculator>(0, SimpleMethodMapperFactory)
    val messageSerializer = messageSerializer(JavaSerializer)
    val address = InetSocketAddress("localhost", 28947)
    val executor = Executors.newCachedThreadPool()
    val server = Server(Service(calculatorId, CalculatorImpl()))
    socketServer(ServerSetup(server, messageSerializer), executor)
        .start(executor, socketBinder(address))
    val client = socketClient(ClientSetup(messageSerializer), socketConnector(address))
    useCalculator(client.proxy(calculatorId))
}
